src/components/fields/ArrayField.js

```diff
@@ -189,36 +176,29 @@ class ArrayField extends Component {
     return itemsSchema.type === "string" && itemsSchema.minLength > 0;
   }
 
-  asyncSetState(state, options={validate: false}) {
-    setState(this, state, () => {
-      this.props.onChange(this.state.items, options);
-    });
-  }
-
   onAddClick = (event) => {
     event.preventDefault();
-    const {items} = this.state;
-    const {schema, registry} = this.props;
+    const {schema, registry, formData} = this.props;
     const {definitions} = registry;
     let itemSchema = schema.items;
     if (isFixedItems(schema) && allowAdditionalItems(schema)) {
       itemSchema = schema.additionalItems;
     }
-    this.asyncSetState({
-      items: items.concat([
-        getDefaultFormState(itemSchema, undefined, definitions)
-      ])
-    });
+    this.props.onChange([
+      ...formData,
+      getDefaultFormState(itemSchema, undefined, definitions)
```

Comment:
I'm not sure whether it is appropriate to do it at the moment because getDefaultFormState is used  by the root Form component as well.